internal/releasesources/release_sources.go

```diff
@@ -0,0 +1,478 @@
+}
+
+// removeReleaseImageDuplicateRecords removes duplcates from release images slice
+func (h *ReleaseSourcesHandler) removeReleaseImageDuplicateRecords(releases []*common.ReleaseImage) []*common.ReleaseImage {
+	resultReleases := []*common.ReleaseImage{}
+	releaseImageReferences := map[string]bool{}
+
+	for _, releaseImage := range releases {
+		if releaseImageReferences[*releaseImage.URL] {
+			continue
+		}
+
+		releaseImageReferences[*releaseImage.URL] = true
+		resultReleases = append(resultReleases, releaseImage)
+	}
+
+	return resultReleases
+}
+
+func (h *ReleaseSourcesHandler) removeReleaseImageDuplicates(releases []*common.ReleaseImage) []*common.ReleaseImage {
+	releases = h.removeReleaseImageDuplicatesFromDifferentChannels(releases)
+	releases = h.removeReleaseImageDuplicateRecords(releases)
+	return releases
+}
+
+// getReleasesWithoutDuplicates removes duplicate release images from different channels
+func (h *ReleaseSourcesHandler) removeReleaseImageDuplicatesFromDifferentChannels(releases []*common.ReleaseImage) []*common.ReleaseImage {
+	resultReleases := []*common.ReleaseImage{}
+	stableReleasesReferences := map[string]bool{}
+
+	for _, release := range releases {
+		if *release.Channel == OpenshiftReleaseChannelStable {
+			stableReleasesReferences[*release.URL] = true
+		}
+	}
+
+	// Add only releases which does not have a stable duplicate
+	for _, release := range releases {
+		if *release.Channel != OpenshiftReleaseChannelStable && stableReleasesReferences[*release.URL] {
+			continue
+		}
+
+		resultReleases = append(resultReleases, release)
+	}
+
+	return resultReleases
+}
+
+func (h *ReleaseSourcesHandler) getDynamicReleaseImages() ([]*common.ReleaseImage, error) {
+	// OCP releases API needs amd64, arm64 instead of x86_64, aarch64 respectively
+	cpuArchMapToAPIArch := map[string]string{
+		common.X86CPUArchitecture:     common.AMD64CPUArchitecture,
+		common.AARCH64CPUArchitecture: common.ARM64CPUArchitecture,
+	}
+
+	// switch back
+	cpuArchMapFromAPIArch := map[string]string{
+		common.AMD64CPUArchitecture: common.X86CPUArchitecture,
+		common.ARM64CPUArchitecture: common.AARCH64CPUArchitecture,
+	}
+
+	releases := []*common.ReleaseImage{}
+	// Aggregate slice of releases from releases API
+	for _, releaseSource := range h.releaseSources {
+		openshiftVersion := *releaseSource.OpenshiftVersion
+		for _, upgradeChannel := range releaseSource.UpgradeChannels {
+			cpuArchitecture := *upgradeChannel.CPUArchitecture
+			for _, channel := range upgradeChannel.Channels {
+				apiCpuArchitecture, shouldSwitch := cpuArchMapToAPIArch[cpuArchitecture]
+				if shouldSwitch {
+					cpuArchitecture = apiCpuArchitecture
+				}
+
+				graph, err := h.releasesClient.GetReleases(channel, openshiftVersion, cpuArchitecture)
+				if err != nil {
+					h.log.WithError(err).Warnf("failed to get releases from api with: %s-%s, %s", channel, openshiftVersion, cpuArchitecture)
+					continue
+				}
+
+				if shouldSwitch {
```

Comment:
Why this shouldSwitch is needed? can't we just get it from map? I mean it is not enough to get cpu arch from mapping dict only once 
https://github.com/openshift/assisted-service/pull/5916/files#diff-22441f7a4340c9547ef0e7b14e79b43caca92e10721eb318513bfb480653aa45R157 ?

Reply:
The API requires some terminology that might be different from ours, for example `x86_64` should be changed to `amd64`. I wanted the initial specification to remain, meaning if one day we will decide to name it `amd64` it will not brake this code